interpreters/wamr/wamr_custom_init.c

```diff
@@ -0,0 +1,657 @@
+/****************************************************************************
+ * apps/interpreters/wamr/wamr_custom_init.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <semaphore.h>
+#include <sys/param.h>
+#include <sys/types.h>
+#include <iconv.h>
+
+#include "wasm_export.h"
+#include "wasm_native.h"
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static sem_t g_aligned_memory_map_sem = SEM_INITIALIZER(1);
+static uintptr_t g_aligned_memory_map[CONFIG_INTERPRETERS_WAMR_LIBC_NUTTX_POSIXMEMALIGN_MAP_SIZE][2] = {{0, 0}};
+
+static bool
+add_to_aligned_map(uintptr_t mapped, uintptr_t raw)
+{
+  int i;
+  bool ret = false;
+
+  sem_wait(&g_aligned_memory_map_sem);
+
+  for (i = 0; i < CONFIG_INTERPRETERS_WAMR_LIBC_NUTTX_POSIXMEMALIGN_MAP_SIZE; i++)
+    {
+      DEBUGASSERT(g_aligned_memory_map[i][0] != mapped);
+
+      if (g_aligned_memory_map[i][0] == 0 && g_aligned_memory_map[i][1] == 0)
+      {
+        g_aligned_memory_map[i][0] = mapped;
+        g_aligned_memory_map[i][1] = raw;
+        ret = true;
+        break;
+      }
+    }
+
+  sem_post(&g_aligned_memory_map_sem);
+  return ret;
+}
+
+static uintptr_t
+remove_from_aligned_map(uintptr_t mapped)
+{
+  int i;
+  uintptr_t ret = 0;
+
+  if (mapped == 0)
+    {
+      return ret;
+    }
+
+  sem_wait(&g_aligned_memory_map_sem);
+
+  for (i = 0; i < CONFIG_INTERPRETERS_WAMR_LIBC_NUTTX_POSIXMEMALIGN_MAP_SIZE; i++)
```

Comment:
Please try to reduce the symbol name, i.e.: just use MEMALIGN instead of POSIXMEMALIGN